### PR TITLE
Pink runtime 1.1

### PR DESCRIPTION
**Pull request by kvinwang**
# Overview
This PR bumps the version of pink-runtime to `1.1`. It also implements the logic related to pink runtime upgrade in pruntime. The newly released pruntime will carry the following pink runtimes:
```
$ ll bin/libpink*
-rwxr-xr-x 1 kvin kvin 6391288 May 23 08:52 bin/libpink.so.1.0*
-rwxrwxr-x 1 kvin kvin 6805272 May 23 08:52 bin/libpink.so.1.1*
```

# Upgrade Steps

1. Upgrade the pruntime of all workers in the cluster to the latest version.
2. Use the owner account to call `System::upgrade_runtime((1, 1))` via TX to upgrade the cluster.
3. (Optional) Call `PhalaPhatContract::set_pink_runtime_version` to set the default runtime version for subsequently created clusters.

# Cautions
- If upgraded to an unknown version number, the pruntime will panic and exit.
- To ensure that pink runtime is deterministic, our newly released pruntime will carry all previously released old versions of `libpink.so.*` without recompiling, unless it's necessary to fix a bug. By coincidence, the first version `libpink.so.1.0` we released has a bug: the `on_genesis` function in the runtime is not called during cluster initialization (this function writes the storage_version of each pallet into storage as a reference for subsequent runtime upgrade migration). This leads to the generation of wrong data during the migration of subsequent runtime upgrades. This PR provides an ad-hoc solution: if the old version is `1.0`, the `on_genesis` is first called to fix the storage before the migration, and then the on_runtime_upgrade is called in the new runtime for migration. But this requires `libpink.so.1.0` to provide an API to execute `on_genesis`, so we need to patch the old version of libpink.so: #1278. The new pruntime release carries the patched `libpink.so.1.0`.

@jasl, I've added the patched and compiled `libpink.so.1.0` to the repository. However, to minimize differences, it would be best to recompile a patched `libpink.so.1.0` (#1278) in the environment where the previous version was compiled and put it into `standalone/pruntime/pink-runtimes/libpink.so.1.0`.

<details>
  <summary><h1>简体中文<h1></summary>

# 概述
这个PR bump pink-runtime的版本号到`1.1`. 并且在pruntime中实现了pink runtime升级相关的逻辑。
新发布的pruntime会同时携带以下pink runtime：
```
$ ll bin/libpink*
-rwxr-xr-x 1 kvin kvin 6391288 May 23 08:52 bin/libpink.so.1.0*
-rwxrwxr-x 1 kvin kvin 6805272 May 23 08:52 bin/libpink.so.1.1*
```

# 升级步骤
1. 将cluster中的所有worker的pruntime都升级到最新版本。
2. 用owner账号通过TX调用`System::upgrade_runtime((1, 1))`实现cluster升级。
3. （可选）调用`PhalaPhatContract::set_pink_runtime_version`设置后续创建cluster得默认runtime版本。

# 注意事项
- 若升级到未知版本号，pruntime将panic退出。
- 为了pink runtime保证deterministic，我们发布的新版本pruntime会携带所有已发行过的旧版`libpink.so.*`不作重新编译。除非需要修复bug。刚好我们发布的第一版`libpink.so.1.0`有一个bug：在cluster初始化时，没有调用runtime里的on_genesis函数（这个函数的内容是将各pallet的storage_version写入storage中，作为后续runtime upgrade的migration参考）。这就导致我们后面升级runtime做migration时会产生错误数据。本PR中对此bug做了adhoc处理：如果旧版本是`1.0`，则在migration之前先在旧runtime上调用`on_genesis`修复storage，然后再在新runtime上调用on_runtime_upgrade做migration。但这要求`libpink.so.1.0`提供一个API来执行`on_genesis`, 因此我们需要给旧版本的libpink.so打上这个补丁：#1278，新发布的pruntime携带打补丁后的`libpink.so.1.0`。

@jasl 我将patch后编译的`libpink.so.1.0`加到了仓库中。但为了尽量降低差异，我们最好在编译前一个发布版本的环境中重新编译一个#1278补丁后的`libpink.so.1.0`放入`standalone/pruntime/pink-runtimes/libpink.so.1.0`.
</details>

### CODE REVIEW

File: crates/pink/runtime/src/runtime.rs [hunk 1]
```diff
@@ -161,3 +162,17 @@ fn detect_parameter_changes() {
...
     ));
 }
+
+/// Call on_genesis for all pallets. This would put the each pallet's storage version into
+/// the frame_support::StorageVersion.
+pub fn on_genesis() {
+    <AllPalletsWithSystem as frame_support::traits::OnGenesis>::on_genesis();
+}
+
+/// Call on_runtime_upgrade for all pallets
```
Comment (h4x3rotab) [inline crates/pink/runtime/src/runtime.rs:172]: Note that although every pallet has an `on_runtime_upgrade` hook for migration, on the upstream, it's preferred to specifically define the runtime level migrations in this way: `type Migrations = (A, B, C)`

The reason is that, pallet defined migrations is implicit. The runtime maintainer may not notice the migration. However, usually migrations are highly depends on the runtime (e.g. Phala vs Khala). So in the end, Parity simply prefer to run migrations only defined in the runtime and it's checked in each release. They have nearly abandoned the pallet-based migration.

Comment (h4x3rotab) [reply crates/pink/runtime/src/runtime.rs:172]: Here's the example:

https://github.com/paritytech/polkadot/blob/851071470841fb49d5ab3fe7d4f43c6cef5e489f/runtime/kusama/src/lib.rs#L1506-L1507

Comment (kvinwang) [reply crates/pink/runtime/src/runtime.rs:172]: Yes, here we actually defined our custom migration: `type Migrations = (Migration<PinkRuntime>, AllPalletsWithSystem);` where `Migration` is `pallet_contracts::Migration`.
So the `Migrations:: on_runtime_upgrade` will first call the explicit migration from `pallet_contracts` and then the hook-based migrations.
The order is following the default substrate node implementation. We have a few pallets in our tiny runtime and only the `pallet_contracts` needs an explicit migration.

**Review by h4x3rotab — APPROVED** [review]
LGTM

**Pull request merged**